File: src/metrics/metrics.cpp
#include "metrics/metrics.hpp"

#include "core/app_state.hpp"

#include <algorithm>
#include <cstdint>

namespace moonai::metrics {

namespace {

int count_active_food(const Food &food) {
  int active_food = 0;
  for (uint8_t active : food.active) {
    active_food += active ? 1 : 0;
  }
  return active_food;
}

} // namespace

void refresh(AppState &state) {
  MetricsSnapshot &metrics = state.metrics;
  metrics.step = state.runtime.step;
  metrics.predator_count = static_cast<int>(state.predator.size());
  metrics.prey_count = static_cast<int>(state.prey.size());
  metrics.active_food = count_active_food(state.food);
  metrics.predator_species = static_cast<int>(state.predator.species.size());
  metrics.prey_species = static_cast<int>(state.prey.species.size());

  // Calculate average energy for predators
  float predator_energy_sum = 0.0f;
  for (float energy : state.predator.energy) {
    predator_energy_sum += energy;
  }
  metrics.avg_predator_energy = 0.0f;
  if (!state.predator.energy.empty()) {
    metrics.avg_predator_energy = predator_energy_sum / static_cast<float>(state.predator.energy.size());
  }

  // Calculate average energy for prey
  float prey_energy_sum = 0.0f;
  for (float energy : state.prey.energy) {
    prey_energy_sum += energy;
  }
  metrics.avg_prey_energy = 0.0f;
  if (!state.prey.energy.empty()) {
    metrics.avg_prey_energy = prey_energy_sum / static_cast<float>(state.prey.energy.size());
  }

  // Calculate average complexity for predators
  float predator_complexity_sum = 0.0f;
  for (const auto &genome : state.predator.genomes) {
    predator_complexity_sum += static_cast<float>(genome.nodes().size() + genome.connections().size());
  }
  metrics.avg_predator_complexity = 0.0f;
  if (!state.predator.genomes.empty()) {
    metrics.avg_predator_complexity = predator_complexity_sum / static_cast<float>(state.predator.genomes.size());
  }

  // Calculate average complexity for prey
  float prey_complexity_sum = 0.0f;
  for (const auto &genome : state.prey.genomes) {
    prey_complexity_sum += static_cast<float>(genome.nodes().size() + genome.connections().size());
  }
  metrics.avg_prey_complexity = 0.0f;
  if (!state.prey.genomes.empty()) {
    metrics.avg_prey_complexity = prey_complexity_sum / static_cast<float>(state.prey.genomes.size());
  }

  // Calculate generation metrics for predators
  int max_pred_gen = 0;
  int64_t pred_gen_sum = 0;
  for (int gen : state.predator.generation) {
    max_pred_gen = std::max(max_pred_gen, gen);
    pred_gen_sum += gen;
  }
  metrics.max_predator_generation = 0;
  metrics.avg_predator_generation = 0.0f;
  if (!state.predator.generation.empty()) {
    metrics.max_predator_generation = max_pred_gen;
    metrics.avg_predator_generation =
        static_cast<float>(pred_gen_sum) / static_cast<float>(state.predator.generation.size());
  }

  // Calculate generation metrics for prey
  int max_prey_gen = 0;
  int64_t prey_gen_sum = 0;
  for (int gen : state.prey.generation) {
    max_prey_gen = std::max(max_prey_gen, gen);
    prey_gen_sum += gen;
  }
  metrics.max_prey_generation = 0;
  metrics.avg_prey_generation = 0.0f;
  if (!state.prey.generation.empty()) {
    metrics.max_prey_generation = max_prey_gen;
    metrics.avg_prey_generation = static_cast<float>(prey_gen_sum) / static_cast<float>(state.prey.generation.size());
  }
}

} // namespace moonai::metrics

File: rust/moonai-metrics/src/metrics.rs
use moonai_core::MetricsSnapshot;
use moonai_evo::Genome;

pub struct AgentMetricsData<'a> {
    pub energy: &'a [f32],
    pub genomes: &'a [Genome],
    pub generation: &'a [i32],
    pub species_count: usize,
}

pub struct FoodMetricsData<'a> {
    pub active: &'a [u8],
}

pub fn refresh_metrics(
    metrics: &mut MetricsSnapshot,
    step: i32,
    predator_data: AgentMetricsData,
    prey_data: AgentMetricsData,
    food_data: FoodMetricsData,
    predator_births: i32,
    prey_births: i32,
    predator_deaths: i32,
    prey_deaths: i32,
) {
    metrics.step = step;
    metrics.predator_count = predator_data.energy.len() as i32;
    metrics.prey_count = prey_data.energy.len() as i32;
    metrics.active_food = count_active_food(food_data.active);
    metrics.predator_species = predator_data.species_count as i32;
    metrics.prey_species = prey_data.species_count as i32;
    metrics.predator_births = predator_births;
    metrics.prey_births = prey_births;
    metrics.predator_deaths = predator_deaths;
    metrics.prey_deaths = prey_deaths;

    metrics.avg_predator_energy = calculate_avg_energy(predator_data.energy);
    metrics.avg_prey_energy = calculate_avg_energy(prey_data.energy);

    metrics.avg_predator_complexity = calculate_avg_complexity(predator_data.genomes);
    metrics.avg_prey_complexity = calculate_avg_complexity(prey_data.genomes);

    let (max_pred_gen, avg_pred_gen) = calculate_generation_metrics(predator_data.generation);
    metrics.max_predator_generation = max_pred_gen;
    metrics.avg_predator_generation = avg_pred_gen;

    let (max_prey_gen, avg_prey_gen) = calculate_generation_metrics(prey_data.generation);
    metrics.max_prey_generation = max_prey_gen;
    metrics.avg_prey_generation = avg_prey_gen;
}

fn count_active_food(active: &[u8]) -> i32 {
    active.iter().filter(|&&a| a != 0).count() as i32
}

fn calculate_avg_energy(energy: &[f32]) -> f32 {
    if energy.is_empty() {
        return 0.0f32;
    }
    let sum: f32 = energy.iter().sum();
    sum / energy.len() as f32
}

fn calculate_avg_complexity(genomes: &[Genome]) -> f32 {
    if genomes.is_empty() {
        return 0.0f32;
    }
    let sum: f32 = genomes
        .iter()
        .map(|g| (g.nodes().len() + g.connections().len()) as f32)
        .sum();
    sum / genomes.len() as f32
}

fn calculate_generation_metrics(generation: &[i32]) -> (i32, f32) {
    if generation.is_empty() {
        return (0, 0.0f32);
    }
    let max_gen = *generation.iter().max().unwrap_or(&0);
    let sum: i64 = generation.iter().map(|&g| g as i64).sum();
    let avg_gen = sum as f32 / generation.len() as f32;
    (max_gen, avg_gen)
}

#[cfg(test)]
mod tests {
    use super::*;
    use moonai_evo::Genome;

    #[test]
    fn test_count_active_food() {
        let active = vec![1u8, 0, 1, 1, 0];
        assert_eq!(count_active_food(&active), 3);
    }

    #[test]
    fn test_calculate_avg_energy_empty() {
        let energy: Vec<f32> = vec![];
        assert_eq!(calculate_avg_energy(&energy), 0.0);
    }

    #[test]
    fn test_calculate_avg_energy() {
        let energy = vec![1.0, 2.0, 3.0];
        assert_eq!(calculate_avg_energy(&energy), 2.0);
    }

    #[test]
    fn test_calculate_avg_complexity_empty() {
        let genomes: Vec<Genome> = vec![];
        assert_eq!(calculate_avg_complexity(&genomes), 0.0);
    }

    #[test]
    fn test_calculate_generation_metrics_empty() {
        let generation: Vec<i32> = vec![];
        let (max, avg) = calculate_generation_metrics(&generation);
        assert_eq!(max, 0);
        assert_eq!(avg, 0.0);
    }

    #[test]
    fn test_calculate_generation_metrics() {
        let generation = vec![1, 2, 3, 4, 5];
        let (max, avg) = calculate_generation_metrics(&generation);
        assert_eq!(max, 5);
        assert_eq!(avg, 3.0);
    }
}